### ISSUE TITLE
Asking for visualization code

### ISSUE THREAD

**Issue by Zero-coder** [Graph-and-Geometric-Learning/MindLLM#1]
thanks for sharing powerful and hopeful work, would you please share the visual code, how to draw a fig like this.

<img width="774" height="649" alt="Image" src="https://github.com/user-attachments/assets/fd222816-722c-439b-9803-1432f1cda730" />

**Comment by Boltzmachine** [comment]
We use [pycortex](https://gallantlab.org/pycortex/). We will update the complete visualization code once we have time.

**Comment by Zero-coder** [comment]
thanks for telling me this， aligado for your contribution to BCI.

**Comment by Zero-coder** [comment]
could you please share it recently, I want to produce it.

**Comment by Zero-coder** [comment]
could you please how you actually draw this, which saved file such as xxx.npy for attention visualization？

**Comment by Boltzmachine** [comment]
Sorry for the late reply. I was busy with some deadlines. 

To get the attention weights, you just need to jump into the place where attentions are computed (before or after the softmax, somewhere around https://github.com/Graph-and-Geometric-Learning/MindLLM/blob/0b7294cd0eedd503eae75b934dff9c0828ad45d5/src/modules/neuro_informed_attn.py#L64),  and save it during the forward pass.

**Comment by Boltzmachine** [comment]
Also according to https://gallantlab.org/pycortex/database.html, you need to download the overlays from https://drive.google.com/drive/folders/1YXAYggtlsfFxMtiqWs0UrMihxI125MHb?usp=drive_link and put them in the pycortex's database

**Comment by Boltzmachine** [comment]
This code is for now for your reference. We will release a final version after we finalize the parameters.

```python
import cortex
import torch
import torch.nn.functional as F
import nibabel as nib
import numpy as np

def remove_unnormal(data):
    q1 = np.percentile(data[data!=0], 25)
    q2 = np.percentile(data[data!=0], 75)
    iqr = q2-q1

    lower_bound = q1-1.5*iqr
    upper_bound = q2+1.5*iqr

    data[(data<lower_bound)|(data>upper_bound)&(data!=0)] = 0

    return data

def map_1d_to_3d(array_1d, mask_3d):
    
    array_1d = np.asarray(array_1d)
    mask_3d = np.asarray(mask_3d, dtype=bool)

    if array_1d.size != np.sum(mask_3d):
        raise ValueError("1D array length must match number of True values in mask")
    
    result = np.zeros(mask_3d.shape)

    result[mask_3d] = array_1d
    
    return result

# get mask
mask = nib.load("roi/subj01/nsdgeneral.nii.gz").get_fdata() > 0
subject = 'subj01'
xfm = 'func1pt8_to_anat0pt8_autoFSbbr'
# get roi
roi_list = ['FFA-1', 'FFA-2', 'OFA', 'EBA', 'PPA', 'EarlyVis', 'OPA']
roi_masks = cortex.utils.get_roi_masks(subject, xfm, roi_list=roi_list, gm_sampler='cortical-conservative', split_lr=False, threshold=None, return_dict=True)
# get attn
attentions = torch.load("attn.pt").cpu().numpy()
attentions = remove_unnormal(attentions)
attentions = F.softmax(torch.Tensor(attentions, device='cpu'), dim=1).numpy()  # could try both w/ or w/o softmax

for i in range(attentions.shape[0]):
    attention = map_1d_to_3d(attentions[i].reshape(-1), mask)
    threshold_high = np.percentile(attention[attention!=0], 95)
    threshold_low = np.percentile(attention[attention!=0], 0)
    pdata = attention.transpose((2,1,0))
    pdata[pdata==0] = np.nan
    vol_data = cortex.Volume(pdata, subject, xfm, cmap='Reds', vmin=threshold_low, vmax=threshold_high)
        
    filename = f'figures/subj01_{i}.png'
    _ = cortex.quickflat.make_png(filename, vol_data, recache=False, with_curvature=False, with_sulci=True, with_label=True, with_colorbar=False, with_rois=True\
                                    , sampler='nearest', depth=0.5, pixelwise=True, thick=1\
                                    , curvature_contrast=0.5, curvature_brightness=0.1, roi_list=roi_list)
```

**Comment by Boltzmachine** [comment]
Close as no further activity.
